### PR TITLE
Switch from alt_bn255 to BLS12-381 and add domain separation

### PR DESCRIPTION
**Pull request by vbuterin**

### CODE REVIEW

**Comment by JustinDrake** [comment]
Suggestion: replace `get_fork_version(state, slot)` with `get_domain(state, slot, domain_type)` to abstract away the logic `get_fork_version(state, slot) * 2**32 + DOMAIN_TYPE` repeated several times.

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -691,10 +698,15 @@ def min_empty_validator(validators: List[ValidatorRecord], current_slot: int):
     return None
 ```
 
+```python
+def get_fork_version(state: State, slot: uint64):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: In Python:

```suggestion
def get_fork_version(state: State, slot: int) -> int:
```

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -893,16 +908,16 @@ Run `exit_validator(data.validator_index, state, block, penalize=False, current_
...
     'vote2_data': AttestationSignedData,
-    'vote2_aggregate_sig': '[uint256]',
+    'vote2_aggregate_sig': '[uint384]',
 }
 ```
 
 Perform the following checks:
 
-* For each `aggregate_sig`, verify that `BLSVerify(pubkey=aggregate_pubkey([validators[i].pubkey for i in aggregate_sig_indices]), msg=vote_data, sig=aggsig)` passes.
+* For each `aggregate_sig`, verify that `BLSVerify(pubkey=aggregate_pubkey([validators[i].pubkey for i in aggregate_sig_indices]), msg=vote_data, sig=aggsig, domain=get_fork_version(state, vote_data.slot) * 2**32 + DOMAIN_ATTESTATION)` passes.
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: `aggsig` -> `vote_aggregate_sig`

```suggestion
* For each `aggregate_sig`, verify that `BLSVerify(pubkey=aggregate_pubkey([validators[i].pubkey for i in aggregate_sig_indices]), msg=vote_data, sig= vote_aggregate_sig, domain=get_fork_version(state, vote_data.slot) * 2**32 + DOMAIN_ATTESTATION)` passes.
```

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -92,6 +92,15 @@ The primary source of load on the beacon chain are "attestations". Attestations
 | `ENTRY` | `0` |
 | `EXIT` | `1` |
 
+**Domains for BLS signatures**
+
+| Name | Value | 
+| - | :-: |
+| `DOMAIN_DEPOSIT` | `0` |
+| `DOMAIN_ATTESTATION` | `1` |
+| `DOMAIN_PROPOSAL` | `2` |
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: @vbuterin I just merged `master` branch (includes `PROPOSER_SLASHING` stuff) into this PR, so you probably want to update using `DOMAIN_PROPOSAL` stuff in this PR?

**Pull request merged**